src/dsql/DsqlCompilerScratch.cpp

```diff
@@ -238,300 +238,174 @@ void DsqlCompilerScratch::putBlrMarkers(ULONG marks)
-		if (field->typeOfTable.object.hasData())
+		if (type.typeOfTable.object.hasData())
 		{
-			if (field->explicitCollation)
-			{
-				if (field->typeOfTable.schema != ddlSchema)
-				{
-					appendUChar(blr_column_name3);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfTable.schema.c_str());
-					appendMetaString(field->typeOfTable.object.c_str());
-					appendMetaString(field->typeOfName.object.c_str());
-					appendUChar(1);
-					appendUShort(field->textType);
-				}
-				else
-				{
-					appendUChar(blr_column_name2);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfTable.object.c_str());
-					appendMetaString(field->typeOfName.object.c_str());
-					appendUShort(field->textType);
-				}
-			}
-			else
-			{
-				if (field->typeOfTable.schema != ddlSchema)
-				{
-					appendUChar(blr_column_name3);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfTable.schema.c_str());
-					appendMetaString(field->typeOfTable.object.c_str());
-					appendMetaString(field->typeOfName.object.c_str());
-					appendUChar(0);
-				}
-				else
-				{
-					appendUChar(blr_column_name);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfTable.object.c_str());
-					appendMetaString(field->typeOfName.object.c_str());
-				}
-			}
+			putFieldName<true>(type, useExplicitCollate);
 		}
 		else
 		{
-			if (field->explicitCollation)
-			{
-				if (field->typeOfName.schema != ddlSchema)
-				{
-					appendUChar(blr_domain_name3);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfName.schema.c_str());
-					appendMetaString(field->typeOfName.object.c_str());
-					appendUChar(1);
-					appendUShort(field->textType);
-				}
-				else
-				{
-					appendUChar(blr_domain_name2);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfName.object.c_str());
-					appendUShort(field->textType);
-				}
-			}
-			else
-			{
-				if (field->typeOfName.schema != ddlSchema)
-				{
-					appendUChar(blr_domain_name3);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfName.schema.c_str());
-					appendMetaString(field->typeOfName.object.c_str());
-					appendUChar(0);
-				}
-				else
-				{
-					appendUChar(blr_domain_name);
-					appendUChar(field->fullDomain ? blr_domain_full : blr_domain_type_of);
-					appendMetaString(field->typeOfName.object.c_str());
-				}
-			}
+			putFieldName<false>(type, useExplicitCollate);
 		}
-
 		return;
 	}
 
-	switch (field->dtype)
+	// Maybe it is possible to use GEN_descriptor here?
+	putFieldType(type, useSubType);
+}
+
+void DsqlCompilerScratch::putFieldType(const TypeClause& type, const bool useSubType)
```

Comment:
I suggested to rename this function to `putDType()` after the original `putDtype()` was made an overloaded `putType()`. The reason is simple - it doesn't care about fields, it only generates BLR for the given data type.